lib/puppet/functions/simplib/dlookup.rb

```diff
@@ -0,0 +1,140 @@
+# A function for performing lookups targeted at ease of use with defined types.
+#
+# Quite often you need to override something in an existing defined type and,
+# presently, you have to do this by creating a resource collector and
+# potentially ending up with unintended side-effects.
+#
+# This function introduces the capability to consistently opt-in to a lookup
+# syntax for overriding all parameters of a given defined type or parameters on
+# a specific instance of a defined type.
+#
+# This calls `simplib::lookup` under the hood after formatting the parameters
+# appropriately but was split out in case the underlying syntax needs to
+# change in the future.
+#
+# There are two ways to call this method and, in both cases, to prevent
+# conflicts with class parameters, the keys in Hiera have been modified to
+# completely avoid items that are allowed Puppet Class names but to also mirror
+# common Puppet nomenclature for working with existing resources.
+#
+# @example Global Options
+#
+#   In this case, you want to set a parameter on *every* instance of your
+#   defined type that is ever called. For example, this may be useful for
+#   setting cipher suites to a modified global default to meet company policy.
+#
+#   This follows the general Puppet nomenclature for class lookups since you
+#   cannot have a class and defined type of the same name.
+#
+#   Function Call:
+#
+#   ```ruby
+#   define mydef::global (
+#     $ssl_version = simplib::dlookup('mydef::global', 'ssl_version', { 'default_value' => 'SSLv3' })
+#   ) { ... }
+#
+#   mydef::global { 'test': }
+#   ```
+#
+#   Example Hieradata:
+#
+#   ```yaml
+#   ---
+#   mydef::global::ssl_version: 'TLS1.2'
+#   ```
+#
+# @example Specific Instance Options
```

Comment:
This is not the API I was envisioning.  This says that when you create the define, you have to specify whether it can be globally overridden or individually overridden .  I was expecting a lookup that would first look for an individual override, then look for a global override and finally look for the value when the define was made.

Reply:
@lnemsick-simp This is basically just a more user friendly way of achieving the same thing without needing to pass around option hashes and bloat the calls even more in the parameter calls (they're getting pretty darn big).